components/molecules/MainSearchBar/MainSearchBar.tsx

```diff
@@ -1,44 +1,189 @@
+      <S.SearchBarForm onSubmit={(e) => onSubmit(e)}>
+        <input
+          type="text"
+          placeholder="tagSearch"
+          onKeyUp={(e: React.KeyboardEvent<HTMLInputElement>) => {
+            setInputValue(e.target.value);
+          }}
+          ref={input}
+        />
+        <button type="button">
+          <Image src={searchIcon} alt="search icon" />
+        </button>
+      </S.SearchBarForm>
+      {inputValue !== '' ? (
+        <S.FilteredTagList>
+          {wholeTagList
+            .filter(({ name }) => name.includes(inputValue))
+            .map(({ id, name, color }) => (
+              <S.FilteredTag
+                key={id}
+                color={color}
+                onClick={() => {
+                  onFilteredTagClick({ id, name, color });
+                  onClick();
+                }}
+              >
+                {name}
+              </S.FilteredTag>
+            ))}
+        </S.FilteredTagList>
+      ) : (
+        <S.FilteredTagList>
+          {wholeTagList.map(({ id, name, color }) => (
+            <S.FilteredTag
+              key={id}
+              color={color}
+              onClick={() => {
+                onFilteredTagClick({ id, name, color });
+                onClick();
+              }}
+            >
+              {name}
+            </S.FilteredTag>
+          ))}
+        </S.FilteredTagList>
+      )}
+    </S.TagSearchBarContainer>
+  );
+}
+
+function TitleSearchBar({
+  handleKeyWord,
+}: {
+  handleKeyWord: (keyWord: string) => void;
+}) {
+  const input = useRef<HTMLInputElement>(null);
+  const onSubmit = (e: React.FormEvent) => {
+    e.preventDefault();
+    // 이 부분에서 도저히 any를 쓰지 않고 어떻게 해야할 지 잘 모르겠습니다 ㅜㅠ
+    const { target }: { target: any } = e;
+    handleKeyWord(target[0].value);
+  };
+  return (
+    <S.SearchBarForm onSubmit={(e) => onSubmit(e)}>
+      <input type="text" placeholder="TitleSearch" ref={input} />
       <button type="button">
         <Image src={searchIcon} alt="search icon" />
       </button>
     </S.SearchBarForm>
   );
 }
 
-export function MainSearchBar() {
+export function MainSearchBar({
+  handlePostFilter,
+  wholeTagList,
+}: IMainSearchBarProps) {
+  const [isTagSearch, setIsTagSearch] = useState(false);
+  const [tagList, setTagList] = useState<ITag[]>([]);
+  const [keyWord, setKeyWord] = useState('');
+
+  const onCheckboxClick = () => {
+    setIsTagSearch((v) => !v);
+  };
+
+  const handleKeyWord = (newKeyWord: string) => {
+    setKeyWord(newKeyWord);
+  };
+
+  const addTagList = (tag: ITag) => {
+    if (!tagList.some((eachTag) => eachTag.id === tag.id)) {
+      const temp = tagList;
+      temp.push(tag);
+      setTagList(() => temp);
+    }
+  };
+
+  const deleteTagList = (tagName: string) => {
+    tagList.map((tag) => {
+      if (tag.name === tagName) {
+        const temp = tagList.filter((each) => each !== tag);
+        setTagList(temp);
+      }
+      return null;
+    });
+  };
```

Comment:
여기에서 태그를 삭제하기 위해 `map`을 사용하셨는데, 단순 순회 용도로 사용하신것 같다는 느낌이 드네요!

그럴땐, for문 또는 forEach, for in, for of 를 사용하는 것이 좋아보입니다.

그리고 테스트를 안해봤지만, 제가 감히 생각하기엔 filter 함수 하나로 처리가 가능할 것 같은데 어떠신가요?

```suggestion
  const deleteTagList = (tagName: string) => {
    setTagList((prev) => prev.filter((each) => each.name !== tagName));
  }
```

Reply:
추가로 첨언드리면, for 문을 리액트에서 사용할 경우 liint가 에러나 경고를 낼 겁니다 :)

Reply:
당신들 그만 빛나라구..!! ✨✨✨✨✨✨
![image](https://user-images.githubusercontent.com/97934878/184272640-a5ad9737-7b2d-4d8a-b8af-83750c8f9821.png)